### PR TITLE
Store Code Metrics

### PR DESCRIPTION
**Pull request by johaneshp**
## 📝 Description
This PR stores metrics produced in individual code analysis to the database, both for llm and non llm.
New table created: llm_code_individual, and non_llm_code_individual. The storing of each metrics allows reusing metrics for future use in case user wants to add more context to the project, or obtain previously analyzed project.
The function store project summary of llm code analysis, and complexity metrics of non llm code analysis.
The get function use classification_id to retrieve both metrics.

> Please include a summary of the change and which issue is fixed.  
> Include relevant motivation and context.  
> List any dependencies that are required for this change.

**Closes:** #189 
---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [x] ♻️ Refactoring
- [ ] ⚡ Performance improvement

---

## 🧪 Testing

> Please describe how you tested this PR (both manually and with tests).  
> Provide instructions so we can reproduce.

- [x] pytest
- [x] Test manually with own file

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [ ] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [x] 🔗 Any dependent changes have been merged and published in downstream modules
- [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

> If applicable, add screenshots to help explain your changes

<details>
<summary>Click to expand screenshots</summary>

Screenshot 2025-11-24 at 8 32 22 PM
Screenshot 2025-11-24 at 8 31 54 PM
Screenshot 2025-11-24 at 8 32 04 PM

</details>

### CODE REVIEW

**Review by taoTimTim — CHANGES_REQUESTED** [review]
Good work.

Please close the issue #189 in this PR.

All tests are passing. 

After running the system, all metrics are being stored correctly in the database.
LLM code:
image

Non-LLM code:
image

I think it would be best to simplify `db/code_metrics.py` by removing all logic from the storage functions so that the file contains only the raw SQL queries. This would make the queries reusable across different parts of the system.

Specifically:
- `store_code_complexity_metrics()` should be split so that all data extraction happens outside the DB layer
- `store_code_llm_metrics()` can also be separated so that existence checks (SELECT metrics_id) are handled in the calling function, not inside the DB query function

This keeps the database layer focused solely on insert/update operations, and improves  testability and reusability.

File: src/db/code_metrics.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    comment_ratio: float,
+    total_functions: int,
+    avg_complexity: float,
+    avg_maintainability: float,
+    functions_needing_refactor: int,
+    high_complexity_files: int,
+    low_maintainability_files: int,
+    radon_details_json: str,
+    lizard_details_json: str,
+    update: bool = False
```
Comment (taoTimTim) [inline src/db/code_metrics.py]: Again, the type of call should be determined in the calling function, not here.
These should be two separate queries so they are reusable.

**Review by taoTimTim — CHANGES_REQUESTED** [review]
This refactor is better, thank you. I still have one concern about maintainability:

Right now the query functions still contain logic that’s specific to this part of the system (they rely on an update flag to decide whether to insert or update). This makes the queries less reusable in the future. If another part of the system wants to insert/update this table, it would have to re-implement its own query functions, because the current versions are tied to a particular workflow.

To make these DB helpers more flexible and reusable, we should:
- Remove all business logic from the query functions. The DB layer should be only CRUD,  meaning each function does exactly one SQL operation: INSERT, UPDATE, GET, EXISTS

- Move the decision-making into `project_analysis.py`. That file should determine whether we need an insert or update by calling the existing *_exists() helper.

This separation keeps the DB helpers generic and makes them usable across the whole system without introducing new attributes.

**Review by taoTimTim — APPROVED** [review]
Thanks for fixing that, Johanes! This makes it much easier to reuse the queries later. All tests are passing. Approved.

**Review by AdaraPutri — CHANGES_REQUESTED** [review]
hey Johanes, thanks for working on this! I tried it with two code projects and the db storage is as expected, all tests are passing and the logic of the table and its query functions is well written. great job! 

I noticed llm_code_individual is only storing the llm-generated project summary but not the llm-generated contribution summary. No changes needed in the flow though, I think there's no need to store the latter because we're most likely going to delete that since we now have the non-llm code skill detector results (which is more reliable).

It is the shortcoming of our current flow (unrelated to this pr) that we have not asked users for manual input of project summaries yet for non llm (which we will definitely need in the resume and portfolio). Since all projects (llm or not) have their complexity data and other metadata stored in `non_llm_code_individual`, and the only llm metric stored is the summaries generated, I recommend we change the table name `llm_code_individual` to be called `code_project_summary`, so future non-llm code project summaries can also be stored in it. and add a field `llm_generated` which is a boolean (True if llm consent given, False otherwise). --> this should be further implemented once non-llm summary prompt feature is added

This means the `insert_code_llm_metrics`, `update_code_llm_metrics`, `insert_code_llm_metrics`, `get_code_llm_metrics`, and `code_llm_metrics_exists` functions in `code_metrics.py` would also need to be renamed.

Another option could be to store the summaries directly to `project_summaries` in the field `summary_text`, which would make the current `llm_code_individual` table unnecessary and redundant. Let me know your thoughts!

Everything else looks good! The test cases cover all important edge cases.

**Comment by johaneshp** [comment]
hey @AdaraPutri ! thanks for the feedback, i decided to remove the storage of project summary of llm since it is already covered in project_summaries table.

**Review by AdaraPutri — APPROVED** [review]
thanks Johanes! this branch is now ready to be merged
